src/porepy/utils/tangential_normal_projection.py

```diff
@@ -17,111 +21,95 @@ class TangentialNormalProjection:
+    """
 
+    def __init__(self, normals: np.ndarray) -> None:
         # Normalize vectors
         normals = normals / np.linalg.norm(normals, axis=0)
 
-        self.num_vecs = normals.shape[1]
-        self.dim = dim
+        self.num_vecs: int = normals.shape[1]
+        """Number of normal vectors represented by this object."""
+        self.dim: int = normals.shape[0]
+        """Dimension of the ambient space."""
 
         # Compute normal and tangential basis
         basis, normal = self._decompose_vector(normals)
 
-        basis = basis.reshape((dim, dim, self.num_vecs))
-        self.tangential_basis = basis[:, :-1, :]
+        basis = basis.reshape((self.dim, self.dim, self.num_vecs))
 
         # The projection is found by inverting the basis vectors
-        self.projection = self._invert_3d_matrix(basis)
+        self._projection = self._invert_3d_matrix(basis)
+        """Projection matrix to the local coordinate system."""
+
         self.normals = normal
+        """Normal vectors, ``shape=(dim, num_vecs)``."""
 
     ## Methods for genertation of projection matrices
 
-    def project_tangential_normal(self, num=None):
+    def project_tangential_normal(self, num: Optional[int] = None) -> sps.spmatrix:
         """Define a projection matrix to decompose a matrix into tangential
         and normal components.
 
-        The intended usage is to decompose a grid-based vector variable into the
-        tangent and normal spaces of the grid, with the tacit understanding that there is
-        a single normal vector shared for all the cells (or faces) in the grid.
+        The intended usage is to decompose a vector defined in the global coordinate
+        sytsem variable into the tangent and normal spaces of a local coordinate system.
 
         The method can also create projection matrix based on unequal normal vectors.
-        One projection will be generated per column in self.normal. To activate
-        this behavior, set num=None.
+        One projection will be generated per column in self.normal. To activate this
+        behavior, set num=None.
 
         Parameters:
-            num (int, optional): Number of (equal) projections to be generated.
-                Will correspond to the number of cells / faces in the grid.
-                The projection matrix will have num * self.dim columns. If not
-                specified (default), one projection will be generated per vector in
-                self.normals.
-                NOTE: If self.num_vecs > 1, but num is not None, only the first
+            num: Number of projections to be generated. Will correspond
+                to the number of cells / faces in the grid. The projection matrix will
+                have ``num * self.dim columns``. If not specified, one
+                projection will be generated per vector in ``self.normals``.
+                NOTE: If ``self.num_vecs > 1``, but num is not None, only the first
```

Comment:
I find this a bit confusing. But I don't know if that means I want it changed.

Reply:
I rewrote the description, things should be clearer now.